### PR TITLE
Timebased flush - Added ability to automatically flush a MemTable when N seconds have elapsed without any new activity

### PR DESCRIPTION
**Pull request by parekuti**
Added ability to automatically flush a MemTable when N seconds have elapsed without any new activity

### CODE REVIEW

File: coordinator/src/main/scala/filodb.coordinator/DatasetCoordinatorActor.scala [hunk 1]
```diff
@@ -22,19 +22,21 @@ object DatasetCoordinatorActor {
...
+    *
+    * @param rows the rows to ingest.  Must have the same schema, for now, as the columns passed in
    *        when the ingestion was set up.
    * @return Ack(seqNo)
    */
   case class NewRows(ackTo: ActorRef, rows: Seq[RowReader], seqNo: Long) extends DSCoordinatorMessage
 
   // The default minimum amt of memory in MB to allow ingesting more data
-  val DefaultMinFreeMb = 512
+  val defaultMinFreeMb = 512
```
Comment (velvia) [inline coordinator/src/main/scala/filodb.coordinator/DatasetCoordinatorActor.scala]: could we keep the uppercase start of the variable name?  That is done intentionally - uppercase variables = constants.

File: coordinator/src/main/scala/filodb.coordinator/DatasetCoordinatorActor.scala [hunk 2]
```diff
@@ -22,19 +22,21 @@ object DatasetCoordinatorActor {
    * Ingests a bunch of rows into this dataset, version's memtable.
    * Automatically checks at the end if the memtable should be flushed, and sends a StartFlush
    * message to itself to initiate a flush.  This is done before the Ack is sent back.
-   * @param rows the rows to ingest.  Must have the same schema, for now, as the columns passed in
+    *
+    * @param rows the rows to ingest.  Must have the same schema, for now, as the columns passed in
```
Comment (velvia) [inline coordinator/src/main/scala/filodb.coordinator/DatasetCoordinatorActor.scala]: No need to indent a space just for two lines right?

File: coordinator/src/main/scala/filodb.coordinator/DatasetCoordinatorActor.scala [hunk 3]
```diff
@@ -134,38 +137,49 @@ private[filodb] class DatasetCoordinatorActor(projection: RichProjection,
   val tempRows = new ArrayBuffer[RowReader]
   val ackTos = new ArrayBuffer[(ActorRef, Long)]
   var mTableWriteTask: Option[Cancellable] = None
+  var mTableFlushTask: Option[Cancellable] = None
 
-  def makeNewTable(): MemTable = new FiloMemTable(projection, config)
+    def makeNewTable(): MemTable = new FiloMemTable(projection, config)
```
Comment (velvia) [inline coordinator/src/main/scala/filodb.coordinator/DatasetCoordinatorActor.scala]: No need to indent here

File: coordinator/src/main/scala/filodb.coordinator/DatasetCoordinatorActor.scala [hunk 4]
```diff
@@ -134,38 +137,49 @@ private[filodb] class DatasetCoordinatorActor(projection: RichProjection,
...
-        // schedule future write to memtable, don't have enough rows yet
-        mTableWriteTask = Some(context.system.scheduler.scheduleOnce(flushInterval, self, MemTableWrite))
-      }
+    var flushedCallbacks: List[ActorRef] = Nil
+
+    private def ingestRows(ackTo: ActorRef, rows: Seq[RowReader], seqNo: Long): Unit = {
+      if (canIngest) {
+
+        if (mTableFlushTask.isDefined) {
+          logger.info("Cancelling scehduled memtable flush task")
```
Comment (velvia) [inline coordinator/src/main/scala/filodb.coordinator/DatasetCoordinatorActor.scala]: These should be at most DEBUG level, not INFO.

**Comment by velvia** [comment]
@parekuti this looks good overall, just some stylistic comments -- especially, could we avoid unnecessary indents in many places?    

Other things:
- if it's possible, could you squash the two commits into one so we don't have the intelliJ files in the git history?  Squash and do a push -f which would overwrite.
- see if you can play with the thread sleep timers so that Travis CI tests pass.

thanks!

**Comment by parekuti** [comment]
As per your suggestions i modified DataSetCoordinator and test spec. It should be working now. Also we might need intellij file changes in github for new Stress-build. So added back to the github.

File: coordinator/src/main/scala/filodb.coordinator/DatasetCoordinatorActor.scala [hunk 5]
```diff
@@ -146,14 +150,24 @@ private[filodb] class DatasetCoordinatorActor(projection: RichProjection,
...
   private def ingestRows(ackTo: ActorRef, rows: Seq[RowReader], seqNo: Long): Unit = {
     if (canIngest) {
+
+      if (mTableFlushTask.isDefined) {
+        logger.debug("Cancelling scehduled memtable flush task when there is a write activity going on.")
+        mTableFlushTask.foreach(_.cancel)
+      }
+
+      mTableFlushTask = Some(context.system.scheduler.scheduleOnce
+      (mTableNoActivityFlushInterval,self, StartFlush(None)))
```
Comment (velvia) [inline coordinator/src/main/scala/filodb.coordinator/DatasetCoordinatorActor.scala]: minor: indent this line, and space before self?

**Comment by parekuti** [comment]
Fixed it now. Please check.

**Comment by velvia** [comment]
@parekuti sorry your latest commit has lots of unnecessary indents again.  Suggest changing IntelliJ indent spacing to 2 spaces.   Also please run `sbt coordinator/test` and check for any scalastyle warnings...

**Pull request merged**